src/zelos/ext/plugins/yarascan/yarascan.py

```diff
@@ -0,0 +1,423 @@
+# Copyright (C) 2020 Zeropoint Dynamics
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+
+# You should have received a copy of the GNU Affero General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+# ======================================================================
+
+import glob
+import os
+import pathlib
+
+from collections import defaultdict
+from io import StringIO
+from itertools import islice
+from pathlib import Path
+from typing import IO, Generator, List
+
+import zelos
+
+from zelos import CommandLineOption, IPlugin, Zelos
+
+
+CommandLineOption(
+    "yara_file",
+    type=str,
+    action="append",
+    help="Scan memory for yara rules in the specified file(s)",
+)
+CommandLineOption(
+    "yara_file_glob",
+    type=str,
+    default=None,
+    help="Scan memory for yara rules in all files specified by the given glob",
+)
+CommandLineOption(
+    "yara_rule",
+    type=str,
+    action="append",
+    help="Scan memory for the specified yara rule string(s).",
+)
+CommandLineOption(
+    "yara_outfile",
+    type=str,
+    default=None,
+    help="Dump matches to the specified YAML file.",
+)
+CommandLineOption(
+    "yara_pid",
+    type=int,
+    default=None,
+    help="Only scan memory in the specified pid, otherwise scan all.",
+)
+CommandLineOption(
+    "yara_memdump",
+    type=str,
+    default=None,
+    help="Dump matching memory regions to file in the specified directory.",
+)
+CommandLineOption(
+    "yara_brief",
+    action="store_true",
+    help="Only dump the list of matching rules, exclude match string details.",
+)
+CommandLineOption(
+    "yara_xrefs",
+    action="store_true",
+    help="Count memory cross-references (pointers) for each match.",
+)
+CommandLineOption(
+    "yara_max",
+    type=int,
+    default=None,
+    help="Maximum number of yara matches per region (default: unlimited).",
+)
+
+
+def _get_scan_regions(
+    memory: zelos.memory.Memory,
+) -> Generator[zelos.MemoryRegion, None, None]:
+    for mr in sorted(memory.get_regions()):
+        # Skip "heap", which is a staging area for zelos, not the user
+        # mode heap.
+        if mr.kind == "heap":
+            continue
+        yield mr
+
+
+def _count_xrefs(
```

Comment:
I think this function is only called within YaraMatch? Might as well put it in there?